Basic support for M185 mouse

doctor64 opened the pull request:
Added basic support for M185 and logs of two hardware versions

jwkokc commented:
Thank you! This action should close my #333 and quite probably the others referenced in that thread.

doctor64 commented:
Fixed connection notification for M185 new
Added support for mouse speed selection (DPI)
Added support for mouse smooth scroll (not usable in xfce)
Add output of button remapping in cli
Updated documentation

Lekensteyn commented:
Hey @doctor64, thanks for your PR!

Would it be possible to squash some changes into logical commits (it could make review easier).

As for the reprogrammable feature, is it actually usable? IIRC enabling the reprog feature would result in no longer functional keys since it depends on the driver (hid-logitech-hidpp kernel module) forwarding the correct keys.

doctor64 commented:
Hello @Lekensteyn, 

Sorry, i'm rather new to git and GitHub. Did you mean to squash all commits in pull request into single one? May be, you can point me to some manual describing this task? Thanks in advance.

Hmm, did you asking about keys reprogramming feature or smooth scrolling feature?
Scrolling is currently unusable, bat can be useful to disable feature if it enabled by some way (looks like sometimes this setting survive hot reboot from win to linux).
Keys remapping is a different story. At least in mouse, using REPROG_CONTROLS_V4, we can do two different things. 
First, we can divert key events from HID events to HID++ events, of course, this have little use in linux at this time. 
Second, device can remap one physical key to another. For example, my M185 mouse have three physical button. Left and right button can be programmed to generate events for left or right buttons, middle button can be assigned to generate middle, left or right buttons. I was tested this functionality using `hidconsole` and it works without any problems. Unfortunately, this functionality not (yet) implemented in Solaar.

Lekensteyn commented:
Normally you have to use (interactive) "git rebase" to combine changes into one, there is the git manual page for that (but you can find other resources). If you have difficulties with that I can do it for you

> First, we can divert key events from HID events to HID++ events, of course, this have little use in linux at this time.

Yes, this would have to be handled by the hid-logitech-hidpp driver. With help of a userspace tool (Solaar?) this could then be reconfigured I guess.

Button remapping can also be done with xinput (and maybe even libinput), I don't think it is worthwhile to implement this in Solaar?

doctor64 commented:
Hello @Lekensteyn 
I squashed changes, hope this help/ Just divided to three commits - to make changes more structured.

About button remapping - logitech docs stated what remapped key should be reverted to default during device reset, but, in my experience, sometimes settings survive reboot from win to linux, and can make quite a mess if you have both remapping active on hardware and software level. 
May be it's just mine personal habit - i'm ambidextrous and usually use two mouses, for left and right hands. 
I'm currently working on implementation of key remapping editor for solaar, but it can be long, because remapping not fit into existing solaar code - you need to send two commands to device, one to map key A to key B, and second to map key B to key A. This not quite a good fit to solar template functions.

Lekensteyn commented:
Thanks, that was exactly what I meant by squashing and made review easier :-)

"Device reset" refers to removing power from the input device (e.g. turning the mouse/keyboard off), not reboot of the host computer.

@jwkokc Can you give this some testing?

jwkokc commented:
I'm working with the version from the Ubuntu repository, which I patched to enable my (new model) M185; haven't done a full recompile of the latest changes. However I can try if you'll advise me how to get full source from github. I do have all the needed tools, since I develop a few tools in Python and older languages, just need the current *.py files with which to test but am not familiar with the layout of github!

doctor64 commented:
Hi @jwkokc ,
You can clone sources from my repository, or i can send archive to you by e-mail. Just unpack it somewhere, exit from your current version and run bin/solaar from directory where you clone/unpack new version.

jwkokc commented:
Just e-mail me the archive -- zip, rar, 7z, or tga formats are all OK -- and I'll give it a whirl.

doctor64 commented:
Sorry, @jwkokc , unable to find your e-mail address on github, so i just drop file here.
[Solaar-m185.zip](https://github.com/pwr/Solaar/files/797541/Solaar-m185.zip)

jwkokc commented:
Got it and will test shortly -- the mouse is on my other machine.

jwkokc commented:
Initial testing -- appears to work properly. Panel icon's pulldown menu shows new item, the M185n, which brings up the display. Added info on display looks good. Tried the "smooth scrolling" switch and it seemed not to have any effect. Pointer-speed slider works well and is a good addition.

The About display still has a copyright notice from four years ago; it should probably be updated to reflect current year.

All in all, looks good! Will leave it in use to see if anything develops over time.

jrbenito commented on lib/logitech_receiver/receiver.py:
```diff
@@ -336,10 +336,14 @@ def __init__(self, handle, device_info):
+			self.max_devices = 6
 
 		if self.max_devices == 6:
 			self.name = 'Unifying Receiver'
```
Since `c534` is not a unifying receiver, maybe a custom `self.name` should be set. (not critical, just a suggestion)

jrbenito commented on lib/logitech_receiver/descriptors.py:
```diff
@@ -213,7 +213,14 @@ def _D(name, codename=None, kind=None, wpid=None, protocol=None, registers=None,
+_D('Wireless Mouse M185 old', codename='M185o', protocol=4.5, wpid='4055',
+				settings=[
+							_FS.lowres_smooth_scroll(),
+				])
```
I do have a M185 (MK270 combo) that is wpid 4038 and less features than yours. See #234. I will PR its descriptor.

doctor64 replied:
Hello @jrbenito 
Thank you for your input. Looks like Logitech massively re-utilize model numbers for different devices, and model name can't be used for device feature set detection. I'm think we should remove settings parameter from descriptors and totally rely on feature auto-detection. But, I remember i encounter some problems trying to use auto-detect for feature_toggle template function.  I should investigate this in details then have some spare time.

jrbenito replied:
Yes, Logitech does indeed re-utilize model numbers a lot :(

Anyway, I suspect that feature auto-detection has its issues too. We have to look carefully to this, meanwhile, let´s add descriptors for everyone. I PR #342  and marked it as dependent on this one since it needs USB nano receiver recognition. Thanks for your effort. 

Now we wait for @Lekensteyn to have some time to merge.

Lekensteyn replied:
Auto-detection requires the device to be on, could that have been the problem?
The WPID is different, that should be good enough to differentiate devices?

Lekensteyn commented on lib/logitech_receiver/special_keys.py:
```diff
@@ -335,8 +335,11 @@
 )
 TASK._fallback = lambda x: 'unknown:%04X' % x
-
+# hidpp 4.5 ifo from https://lekensteyn.nl/files/logitech/x1b04_specialkeysmsebuttons.html
```
nit: typo

doctor64 commented:
@Lekensteyn I commit requested changes. Should I squash it into previous commit?

jrbenito commented:
This fix #318

doctor64 commented:
@Lekensteyn PR updated, commits squashed.
I use different codenames because i have both mouses and wish to easy recognize which one can report battery status and which not. It's purely cosmetic difference, but I'm think can help to answer "Why my M185 not report battery status?" If you wish to change to official model names - it can easy be done.

Lekensteyn commented:
> I use different codenames because i have both mouses and wish to easy recognize which one can report battery status and which not

Shouldn't both mice be able to report their battery status? Based on your "new" txt file it seems not to be the case? Any idea whether the Windows software reports the mouse status?

doctor64 commented:
@Lekensteyn As you can see, "new" M185 mouse (wpid 4054) doesn't have feature BATTERY STATUS, and yes, windows logitech software also doesn't report battery status on this device.

doctor64 commented:
@Lekensteyn Here is two screen shots from windows Logitech Options software, using two different M185 mouses. As you see, only one display battery status.
![batt](https://cloud.githubusercontent.com/assets/25585116/24309379/4d3a30d4-10d4-11e7-82c9-16fc01a4c0ff.png)
![no_batt](https://cloud.githubusercontent.com/assets/25585116/24309383/5263de52-10d4-11e7-822d-a73b5cd0f870.png)

Lekensteyn commented:
The "new" mouse has an older firmware version than the "old" one. Is there a firmware update available for your "new" mouse? (I am still waiting for further info from Logitech about this btw.)

doctor64 commented:
@Lekensteyn I can't find firmware updates for M185 on Logitech site. "New" and "old" names based on dates of purchase and part numbers - P/N: 810-005238 for new and P/N: 810-003496 for old one.

Lekensteyn commented on lib/logitech_receiver/settings_templates.py:
```diff
@@ -250,6 +256,15 @@ def _feature_adjustable_dpi():
+					write_function_id=0x10,
+					bytes_count=2,
+					label=_POINTER_SPEED[1], description=_POINTER_SPEED[2],
+					device_kind=_DK.mouse)
```
Hey, based on these docs (https://lekensteyn.nl/files/logitech/x2205_pointermotionscaling.html), wouldn't a range from 0x0010 to 0x1000 be better? (What happens if you try those settings?)

doctor64 replied:
Hi @Lekensteyn!
Range from 0x0010 to 0x1000 works, but GUI becomes almost unusable. Good value for pointer_speed on my system is around 256 (0x100) and its almost impossible to move slider to acceptable value - mouse sensitivity become too big.

doctor64 replied:
@Lekensteyn current range for pointer_speed (0x002e, 0x1ff) taken from USB trace on Windows system, using Logitech Options software. As for me, this range allow to use Solaar GUI control without problems and good enough for users.

jrbenito replied:
Agreed

doctor64 commented:
@Lekensteyn  This was purely cosmetics, to fix warning like `14:11:07,689  WARNING [ReceiverListener:hidraw1] logitech_receiver.notifications: 2: connection notification with unknown protocol 0A: Notification(2,41,0A,125440)`
Protocol named M185 because I see such notifications only on M185 devices.
So, I hope, we can finally merge this PR

Lekensteyn commented:
@jrbenito Can you do a final review?

jrbenito commented:
@Lekensteyn ,

I did check code and tested against my M185. Seems good to go.

doctor64 commented:
Sorry, do we plan to merge this PR?

Lekensteyn commented:
Thanks, sorry for the delay

cement-head commented:
Hi All,

  How do I apply this patch for the M185 mouse?

Thanks

cement-head commented:
I built some debs for this from yesterday's GIT: https://drive.google.com/drive/folders/1lUJ6omeYNnPfhydCL1jYpBNTYLfY_mip
